src/Benchmarks/Middleware/JsonMiddleware.cs

```diff
@@ -40,11 +37,8 @@ public Task Invoke(HttpContext httpContext)
                     syncIOFeature.AllowSynchronousIO = true;
                 }
 
-                _writer ??= new Utf8JsonWriter(httpContext.Response.Body);
-                _writer.Reset(httpContext.Response.Body);
-
-                JsonSerializer.Serialize<JsonMessage>(_writer, new JsonMessage { message = "Hello, World!" }, CustomJsonContext.Default.JsonMessage);
-                return Task.CompletedTask;
+                return JsonSerializer.SerializeAsync<JsonMessage>(httpContext.Response.Body, new JsonMessage { message = "Hello, World!" },
```

Comment:
Remove `.AllowSynchronousIO = true` ?

Reply:
Maybe, forgot pushing to the branch would update a PR 😄 
Still need to test and see if the recent STJ change did anything

Reply:
@eiriktsarpalis, I thought https://github.com/dotnet/runtime/pull/78646 was supposed to allow source-gen code to be used with `SerializeAsync`? But when trying I get the exception shown in https://github.com/dotnet/runtime/issues/71933, without creating a combined context. Are we using it wrong here, or is this another version of the same bug and source-gen still isn't usable with `SerializeAsync`?

Reply:
Removing the `GenerationMode = JsonSourceGenerationMode.Serialization` argument made it work. Is that required for `SerializeAsync`?

Reply:
> Removing the GenerationMode = JsonSourceGenerationMode.Serialization argument made it work. Is that required for SerializeAsync?

Streaming serialization requires metadata generation. The change in dotnet/runtime#78646 simply introduces a heuristic that eventually falls back to fast-path serialization provided that the JSON payloads are small enough.

Reply:
@BrennanConroy any reason we can't simply use the `WriteAsJsonAsync` extension methods here, passing in the options instance from DI resolved from the middleware constructor (to avoid request services)? I assume you'd have to add the serialization context to the options during DI setup too. Would make this slightly more "real world" as it'd use JSON web defaults then.

Reply:
Took forever, but I got it working. One unfortunate thing you have to do to make it work using `WriteAsJsonAsync` is to set `JsonSerialierOptions.Encoder` to `null`.

Reply:
You going to push the update so we can see? :smile:

Reply:
Enjoy